library/core/src/pin.rs

```diff
@@ -110,79 +222,223 @@
+//!     succ: *mut Node,
+//!     data: Data,
+//!     /// `Node: Unpin` because `List` expects `Pin`ned pointers to them to remain in place.
+//!     _pin: PhantomPinned,
+//! }
 //!
-//! If your type uses pinning (such as the two examples above), you have to be careful
-//! when implementing [`Drop`][Drop]. The [`drop`] function takes <code>[&mut] self</code>, but this
-//! is called *even if your type was previously pinned*! It is as if the
-//! compiler automatically called [`Pin::get_unchecked_mut`].
+//! impl Drop for Node {
+//!     /// Remove pointers to `self`, allowing reuse of this memory without clients seeing garbage.
+//!     fn drop(&mut self) {
+//!         if self.pred.is_null() || self.succ.is_null() {
+//!             // Not included: code to remove `self` if it is the head or tail of the list.
+//!             return;
+//!         }
 //!
-//! This can never cause a problem in safe code because implementing a type that
-//! relies on pinning requires unsafe code, but be aware that deciding to make
-//! use of pinning in your type (for example by implementing some operation on
-//! <code>[Pin]<[&]Self></code> or <code>[Pin]<[&mut] Self></code>) has consequences for your
-//! [`Drop`][Drop]implementation as well: if an element of your type could have been pinned,
-//! you must treat [`Drop`][Drop] as implicitly taking <code>[Pin]<[&mut] Self></code>.
+//!         unsafe {
+//!             (*self.pred).succ = self.succ;
+//!             (*self.succ).pred = self.pred;
+//!         }
+//!     }
+//! }
+//! ```
 //!
-//! For example, you could implement [`Drop`][Drop] as follows:
+//! For this to work, a [`drop`-related guarantee][drop-guarantee] is required. If a node could
+//! be deallocated or otherwise invalidated without calling [`drop`], the pointers into it from its
+//! neighboring elements would become invalid, which would break the data structure.
+//!
+//! [`List`] itself is *not* address-sensitive.
+//!
+//! # Subtle details
+//! [subtle-details]: #subtle-details
+//!
+//! [`List::append`] above relies on both of [`Pin<P>`]'s guarantees:
+//!
+//! 1.  *Address Stability.* If [`unsafe`] code witnesses any [`p: Pin<P>`][Pin] at any time then
+//!     it may assume that `p.as_ref().get_ref() as *const _` will remain valid, pointing to the
+//!     same object until the end of that object's lifetime.
+//! 2.  *Notice of Destruction.* If `x: T` was ever reachable through any [`Pin<P>`] type, its
+//!     destructor must be run (until it either returns or panics) before `x`'s storage can be
+//!     overwritten. The "until further notice" in (1) includes this mandatory destruction. This is
+//!     often called the "[`Drop`] guarantee".
+//!
+//! ## Address Stability
+//! [address-stability]: #address-stability
+//!
+//! The precise meaning of "address stability" is subtle, because "the same object" is not well-defined.
+//! It is easiest to reason about it in terms of *visibility of mutations*. If [`unsafe`] code mutates
```

Comment:
FWIW, I think this is best described in terms of which invariants / properties must be satisfied by the data. That's how a mathematical formal treatment of pinning would work (or at least I don't see how else it would work, and I know this way works). The key difference between the invariant of a 'regular' Rust value and the invariant of a value in its address-sensitive state is that the latter invariant *can depend on the address*. The regular invariant satisfies the property that it keeps holding when the same sequence of bytes is put at a different address in memory; the address-sensitive invariant does not have that property.

I don't know what 'visibility' is supposed to mean.